lib/logstash/outputs/elasticsearch.rb

```diff
@@ -484,30 +484,69 @@ def receive(event)
   end # def receive
 
   public
+  def submit(actions)
+    resp = submit_bulk_request(actions)
+    handle_bulk_errors(resp)
+    resp
+  end
+
+  private
+  # Return the type of retry this status requires
+  def bin_for_status(status)
+    if SUCCESS_OCDES.include?(status)
+      :success
+    elsif RETRYABLE_CODES.include?(status)
+      :retryable
+    else
+      :fatal
+    end
+  end
+
+  private
+  # Threadsafe (via mutex) way to issue a bulk request of actions
   # synchronize the @current_client.bulk call to avoid concurrency/thread safety issues with the
   # # client libraries which might not be thread safe. the submit method can be called from both the
   # # Stud::Buffer flush thread and from our own retry thread.
-  def submit(actions)
+  def submit_bulk_request(actions)
+    # Hashify the event for ES API
     es_actions = actions.map { |a, doc, event| [a, doc, event.to_hash] }
+
+    submit_bulk_request(actions)
     @submit_mutex.lock
     begin
       bulk_response = @current_client.bulk(es_actions)
     ensure
       @submit_mutex.unlock
     end
-    if bulk_response["errors"]
-      actions_with_responses = actions.zip(bulk_response['statuses'])
-      actions_to_retry = []
-      actions_with_responses.each do |action, resp_code|
-        if RETRYABLE_CODES.include?(resp_code)
-          @logger.warn "retrying failed action with response code: #{resp_code}"
-          actions_to_retry << action
-        elsif not SUCCESS_CODES.include?(resp_code)
-          @logger.warn "failed action with response of #{resp_code}, dropping action: #{action}"
-        end
-      end
-      retry_push(actions_to_retry) unless actions_to_retry.empty?
-    end
+    bulk_response
+  end
+
+  private
+  def handle_bulk_errors(bulk_response)
+    return unless errors
+
+    retryable, fatal = group_statuses(actions,bulk_response['statuses']).values_at(:retryable,:fatal)
```

Comment:
@talevy I've been forever corrupted by clojure. Must.... destructure... all the things....